Can't replace default MongoClient with @Replaces

I'm trying to create an alternative `MongoClient` factory for a Kotlin project (to use [kmongo](https://litote.org/kmongo/)).  I tried this first using `@Named` as a qualifier (see #549 -- works in Java, fails in Kotlin).

I then tried using the `@Replaces` annotation in my `MongoClient` factory:

```kotlin
@Requirements(
        Requires(classes = arrayOf(MongoClient::class)),
        Requires(beans = arrayOf(DefaultMongoConfiguration::class)))
@Factory
class MyMongoClientFactory {

    companion object: KLogging()

    @Bean(preDestroy = "close")
    @Singleton
    @Replaces(MongoClient::class)
    fun myMongoClient(configuration: DefaultMongoConfiguration): MongoClient {
        logger.info { "===== creating myMongoClient" }
        return MongoClient(configuration.buildURI())
    }
}
```

Injection into controller:
```kotlin
    @Inject
    private lateinit var mongoClient: MongoClient
```

The Java code is very similar...except...well, Java.  The Java and Kotlin code both fail with the same `NoSuchBeanException`.  Here's a stacktrace (nearly identical in both Java and Kotlin) with debug logging before the error:

```
21:51:04.085 [pool-1-thread-5] DEBUG i.m.context.DefaultBeanContext - Finding candidate beans for type: class com.mongodb.MongoClient
21:51:04.086 [pool-1-thread-5] DEBUG i.m.context.DefaultBeanContext - Finding candidate beans for type: class io.micronaut.configuration.mongo.reactive.DefaultMongoConfiguration
21:51:04.087 [pool-1-thread-5] DEBUG i.m.context.DefaultBeanContext - Resolved bean candidates [Definition: io.micronaut.configuration.mongo.reactive.DefaultMongoConfiguration] for type: class io.micronaut.configuration.mongo.reactive.DefaultMongoConfiguration
21:51:04.087 [pool-1-thread-5] DEBUG i.m.context.DefaultBeanContext - Finalized bean definitions candidates: [Definition: io.micronaut.configuration.mongo.reactive.DefaultMongoConfiguration]
21:51:04.087 [pool-1-thread-5] DEBUG i.m.context.DefaultBeanContext - Found concrete candidate [Definition: io.micronaut.configuration.mongo.reactive.DefaultMongoConfiguration] for type: io.micronaut.configuration.mongo.reactive.DefaultMongoConfiguration 
21:51:04.087 [pool-1-thread-5] DEBUG i.m.context.DefaultBeanContext - Bean [class com.mongodb.MongoClient] replaces existing bean of type [class com.mongodb.MongoClient]
21:51:04.088 [pool-1-thread-5] DEBUG i.m.context.DefaultBeanContext - Resolved bean candidates [] for type: class com.mongodb.MongoClient
21:51:04.091 [nioEventLoopGroup-1-2] DEBUG i.m.context.DefaultBeanContext - Finding candidate beans for type: interface io.micronaut.http.server.exceptions.ExceptionHandler
21:51:04.093 [nioEventLoopGroup-1-2] DEBUG i.m.context.DefaultBeanContext - Resolved bean candidates [Definition: io.micronaut.http.server.exceptions.ContentLengthExceededHandler, Definition: io.micronaut.http.server.exceptions.JsonExceptionHandler, Definition: io.micronaut.http.server.exceptions.UnsatisfiedArgumentHandler, Definition: io.micronaut.http.server.exceptions.HttpStatusHandler, Definition: io.micronaut.http.server.exceptions.ConversionErrorHandler, Definition: io.micronaut.http.server.netty.converters.UnsatisfiedRouteHandler, Definition: io.micronaut.http.server.netty.converters.DuplicateRouteHandler, Definition: io.micronaut.validation.exceptions.ConstraintExceptionHandler] for type: interface io.micronaut.http.server.exceptions.ExceptionHandler
21:51:04.093 [nioEventLoopGroup-1-2] DEBUG i.m.context.DefaultBeanContext - Finalized bean definitions candidates: [Definition: io.micronaut.http.server.exceptions.ContentLengthExceededHandler, Definition: io.micronaut.http.server.exceptions.JsonExceptionHandler, Definition: io.micronaut.http.server.exceptions.UnsatisfiedArgumentHandler, Definition: io.micronaut.http.server.exceptions.HttpStatusHandler, Definition: io.micronaut.http.server.exceptions.ConversionErrorHandler, Definition: io.micronaut.http.server.netty.converters.UnsatisfiedRouteHandler, Definition: io.micronaut.http.server.netty.converters.DuplicateRouteHandler, Definition: io.micronaut.validation.exceptions.ConstraintExceptionHandler]
21:51:04.093 [nioEventLoopGroup-1-2] DEBUG i.m.context.DefaultBeanContext - Qualifying bean [io.micronaut.http.server.exceptions.ExceptionHandler] for qualifier: <DependencyInjectionException,Object> 
21:51:04.094 [nioEventLoopGroup-1-2] DEBUG i.m.context.DefaultBeanContext - No qualifying beans of type [io.micronaut.http.server.exceptions.ExceptionHandler] found for qualifier: <DependencyInjectionException,Object> 

21:32:09.065 [nioEventLoopGroup-1-2] ERROR i.m.h.s.netty.RoutingInBoundHandler - Unexpected error occurred: Failed to inject value for field [mongoClient] of class: example.HelloController

Path Taken: HelloController.mongoClient
io.micronaut.context.exceptions.DependencyInjectionException: Failed to inject value for field [mongoClient] of class: example.HelloController

Path Taken: HelloController.mongoClient
	at io.micronaut.context.AbstractBeanDefinition.getBeanForField(AbstractBeanDefinition.java:1294)
	at io.micronaut.context.AbstractBeanDefinition.injectBeanField(AbstractBeanDefinition.java:691)
	at example.$HelloControllerDefinition.injectBean(Unknown Source)
	at example.$HelloControllerDefinition.build(Unknown Source)
	at io.micronaut.context.DefaultBeanContext.doCreateBean(DefaultBeanContext.java:1252)
	at io.micronaut.context.DefaultBeanContext.createAndRegisterSingleton(DefaultBeanContext.java:1677)
	at io.micronaut.context.DefaultBeanContext.getBeanForDefinition(DefaultBeanContext.java:1447)
	at io.micronaut.context.DefaultBeanContext.getBeanInternal(DefaultBeanContext.java:1427)
	at io.micronaut.context.DefaultBeanContext.getBean(DefaultBeanContext.java:493)
	at io.micronaut.context.DefaultBeanContext$BeanExecutionHandle.invoke(DefaultBeanContext.java:2119)
	at io.micronaut.web.router.AbstractRouteMatch.execute(AbstractRouteMatch.java:290)
	at io.micronaut.web.router.RouteMatch.execute(RouteMatch.java:123)
	at io.micronaut.http.server.netty.RoutingInBoundHandler.lambda$buildResultEmitter$12(RoutingInBoundHandler.java:1175)
	at io.reactivex.internal.operators.flowable.FlowableCreate.subscribeActual(FlowableCreate.java:72)
	at io.reactivex.Flowable.subscribe(Flowable.java:14409)
	at io.reactivex.internal.operators.flowable.FlowableMap.subscribeActual(FlowableMap.java:38)
	at io.reactivex.Flowable.subscribe(Flowable.java:14409)
	at io.reactivex.internal.operators.flowable.FlowableSwitchIfEmpty.subscribeActual(FlowableSwitchIfEmpty.java:32)
	at io.reactivex.Flowable.subscribe(Flowable.java:14409)
	at io.reactivex.Flowable.subscribe(Flowable.java:14359)
	at io.micronaut.http.context.ServerRequestTracingPublisher.lambda$subscribe$0(ServerRequestTracingPublisher.java:53)
	at io.micronaut.http.context.ServerRequestContext.with(ServerRequestContext.java:53)
	at io.micronaut.http.context.ServerRequestTracingPublisher.subscribe(ServerRequestTracingPublisher.java:53)
	at io.reactivex.internal.operators.flowable.FlowableFromPublisher.subscribeActual(FlowableFromPublisher.java:29)
	at io.reactivex.Flowable.subscribe(Flowable.java:14409)
	at io.reactivex.Flowable.subscribe(Flowable.java:14356)
	at io.reactivex.internal.operators.flowable.FlowableSubscribeOn$SubscribeOnSubscriber.run(FlowableSubscribeOn.java:82)
	at io.reactivex.internal.schedulers.ExecutorScheduler$ExecutorWorker$BooleanRunnable.run(ExecutorScheduler.java:261)
	at io.reactivex.internal.schedulers.ExecutorScheduler$ExecutorWorker.run(ExecutorScheduler.java:226)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: io.micronaut.context.exceptions.NoSuchBeanException: No bean of type [com.mongodb.MongoClient] exists. If you are using Java or Kotlin make sure you have enabled annotation processing.
	at io.micronaut.context.DefaultBeanContext.getBeanInternal(DefaultBeanContext.java:1433)
	at io.micronaut.context.DefaultBeanContext.getBean(DefaultBeanContext.java:852)
	at io.micronaut.context.AbstractBeanDefinition.getBeanForField(AbstractBeanDefinition.java:1286)
	... 31 common frames omitted
```

### Task List

- [x] Steps to reproduce provided
- [x] Stacktrace (if present) provided
- [x] Example that reproduces the problem uploaded to Github
- [x] Full description of the issue provided (see below)

### Steps to Reproduce

I modified the `hello-world-java` and `hello-world-kotlin` examples to reproduce the problem.

1. Clone [my fork of micronaut-examples](https://github.com/mmindenhall/micronaut-examples)
2. `git checkout replace-mongodb-client`
3. `cd hello-world-java`
1. `./gradlew run`
1.  GET http://localhost:8080/hello/dude in browser or http client
    1.  `NoSuchBeanException` is thrown
1. `cd ../hello-world-kotlin`
1. `./gradlew run`
1.  GET http://localhost:8080/hello/dude in browser or http client
    1.  `NoSuchBeanException` is thrown

### Expected Behaviour

The `MongoClient` produced from `DefaultMongoClientFactory` is replaced (in both Java and Kotlin).

### Actual Behaviour

`NoSuchBeanException` in both Java and Kotlin.

### Environment Information

- **Operating System**: Mac OS 10.13.6
- **Micronaut Version:** 1.0.0.M4 _and_ 1.0.0.BUILD-SNAPSHOT
- **JDK Version:** 1.8.0_152

### Example Application

See steps to reproduce above.

--- a/configurations/hibernate-jpa/src/main/java/io/micronaut/configuration/hibernate/jpa/HibernateTransactionManagerFactory.java
+++ b/configurations/hibernate-jpa/src/main/java/io/micronaut/configuration/hibernate/jpa/HibernateTransactionManagerFactory.java
@@ -32,7 +32,7 @@ import javax.sql.DataSource;
  */
 @Factory
 @Requires(classes = HibernateTransactionManager.class)
-@Replaces(DataSourceTransactionManagerFactory.class)
+@Replaces(factory = DataSourceTransactionManagerFactory.class)
 public class HibernateTransactionManagerFactory {
 
     /**

--- a/inject-java/src/test/groovy/io/micronaut/inject/qualifiers/replaces/C.java
+++ b/inject-java/src/test/groovy/io/micronaut/inject/qualifiers/replaces/C.java
@@ -1,0 +1,4 @@
+package io.micronaut.inject.qualifiers.replaces;
+
+public interface C {
+}

--- a/inject-java/src/test/groovy/io/micronaut/inject/qualifiers/replaces/C1.java
+++ b/inject-java/src/test/groovy/io/micronaut/inject/qualifiers/replaces/C1.java
@@ -1,0 +1,4 @@
+package io.micronaut.inject.qualifiers.replaces;
+
+public class C1 implements C {
+}

--- a/inject-java/src/test/groovy/io/micronaut/inject/qualifiers/replaces/C2.java
+++ b/inject-java/src/test/groovy/io/micronaut/inject/qualifiers/replaces/C2.java
@@ -1,0 +1,4 @@
+package io.micronaut.inject.qualifiers.replaces;
+
+public class C2 implements C {
+}

--- a/inject-java/src/test/groovy/io/micronaut/inject/qualifiers/replaces/CFactory.java
+++ b/inject-java/src/test/groovy/io/micronaut/inject/qualifiers/replaces/CFactory.java
@@ -1,0 +1,14 @@
+package io.micronaut.inject.qualifiers.replaces;
+
+import io.micronaut.context.annotation.Factory;
+
+import javax.inject.Singleton;
+
+@Factory
+public class CFactory {
+
+    @Singleton
+    C getC() {
+        return new C1();
+    }
+}

--- a/inject-java/src/test/groovy/io/micronaut/inject/qualifiers/replaces/CFactoryReplacement.java
+++ b/inject-java/src/test/groovy/io/micronaut/inject/qualifiers/replaces/CFactoryReplacement.java
@@ -1,0 +1,16 @@
+package io.micronaut.inject.qualifiers.replaces;
+
+import io.micronaut.context.annotation.Factory;
+import io.micronaut.context.annotation.Replaces;
+
+import javax.inject.Singleton;
+
+@Factory
+@Replaces(factory = CFactory.class)
+public class CFactoryReplacement {
+
+    @Singleton
+    C getC() {
+        return new C2();
+    }
+}

--- a/inject-java/src/test/groovy/io/micronaut/inject/qualifiers/replaces/D.java
+++ b/inject-java/src/test/groovy/io/micronaut/inject/qualifiers/replaces/D.java
@@ -1,0 +1,4 @@
+package io.micronaut.inject.qualifiers.replaces;
+
+public interface D {
+}

--- a/inject-java/src/test/groovy/io/micronaut/inject/qualifiers/replaces/D1.java
+++ b/inject-java/src/test/groovy/io/micronaut/inject/qualifiers/replaces/D1.java
@@ -1,0 +1,4 @@
+package io.micronaut.inject.qualifiers.replaces;
+
+public class D1 implements D {
+}

--- a/inject-java/src/test/groovy/io/micronaut/inject/qualifiers/replaces/D2.java
+++ b/inject-java/src/test/groovy/io/micronaut/inject/qualifiers/replaces/D2.java
@@ -1,0 +1,4 @@
+package io.micronaut.inject.qualifiers.replaces;
+
+public class D2 implements D {
+}

--- a/inject-java/src/test/groovy/io/micronaut/inject/qualifiers/replaces/DFactory.java
+++ b/inject-java/src/test/groovy/io/micronaut/inject/qualifiers/replaces/DFactory.java
@@ -1,0 +1,15 @@
+package io.micronaut.inject.qualifiers.replaces;
+
+import io.micronaut.context.annotation.Factory;
+
+import javax.inject.Singleton;
+
+@Factory
+public class DFactory {
+
+    @Singleton
+    D getD() {
+        return new D1();
+    }
+
+}

--- a/inject-java/src/test/groovy/io/micronaut/inject/qualifiers/replaces/DFactoryMethodReplacement.java
+++ b/inject-java/src/test/groovy/io/micronaut/inject/qualifiers/replaces/DFactoryMethodReplacement.java
@@ -1,0 +1,16 @@
+package io.micronaut.inject.qualifiers.replaces;
+
+import io.micronaut.context.annotation.Factory;
+import io.micronaut.context.annotation.Replaces;
+
+import javax.inject.Singleton;
+
+@Factory
+public class DFactoryMethodReplacement {
+
+    @Singleton
+    @Replaces(value = D.class, factory = DFactory.class)
+    D getD() {
+        return new D2();
+    }
+}

--- a/inject-java/src/test/groovy/io/micronaut/inject/qualifiers/replaces/ReplacesSpec.groovy
+++ b/inject-java/src/test/groovy/io/micronaut/inject/qualifiers/replaces/ReplacesSpec.groovy
@@ -70,4 +70,32 @@ class ReplacesSpec extends Specification {
         cleanup:
         ctx.close()
     }
+
+    void "test replacing an entire factory"() {
+        given:
+        def ctx = ApplicationContext.run()
+
+        when:
+        C c = ctx.getBean(C)
+
+        then:
+        c instanceof C2
+
+        cleanup:
+        ctx.close()
+    }
+
+    void "test replacing a factory method"() {
+        given:
+        def ctx = ApplicationContext.run()
+
+        when:
+        D d = ctx.getBean(D)
+
+        then:
+        d instanceof D2
+
+        cleanup:
+        ctx.close()
+    }
 }

--- a/inject-java/src/test/java/io/micronaut/docs/replaces/BookFactory.java
+++ b/inject-java/src/test/java/io/micronaut/docs/replaces/BookFactory.java
@@ -1,0 +1,22 @@
+package io.micronaut.docs.replaces;
+
+import io.micronaut.context.annotation.Factory;
+import io.micronaut.docs.requires.Book;
+
+import javax.inject.Singleton;
+
+// tag::class[]
+@Factory
+public class BookFactory {
+
+    @Singleton
+    Book novel() {
+        return new Book("A Great Novel");
+    }
+
+    @Singleton
+    TextBook textBook() {
+        return new TextBook("Learning 101");
+    }
+}
+// end::class[]

--- a/inject-java/src/test/java/io/micronaut/docs/replaces/CustomBookFactory.java
+++ b/inject-java/src/test/java/io/micronaut/docs/replaces/CustomBookFactory.java
@@ -1,0 +1,20 @@
+package io.micronaut.docs.replaces;
+
+import io.micronaut.context.annotation.Factory;
+import io.micronaut.context.annotation.Replaces;
+import io.micronaut.docs.requires.Book;
+
+import javax.inject.Singleton;
+
+
+// tag::class[]
+@Factory
+@Replaces(factory = BookFactory.class)
+public class CustomBookFactory {
+
+    @Singleton
+    Book otherNovel() {
+        return new Book("An OK Novel");
+    }
+}
+// end::class[]

--- a/inject-java/src/test/java/io/micronaut/docs/replaces/MockBookService.java
+++ b/inject-java/src/test/java/io/micronaut/docs/replaces/MockBookService.java
@@ -31,7 +31,6 @@ import java.util.Map;
  */
 // tag::class[]
 @Replaces(JdbcBookService.class) // <1>
-@Requires(env = Environment.TEST) // <2>
 @Singleton
 public class MockBookService implements BookService {
 

--- a/inject-java/src/test/java/io/micronaut/docs/replaces/ReplacesTest.java
+++ b/inject-java/src/test/java/io/micronaut/docs/replaces/ReplacesTest.java
@@ -17,6 +17,7 @@
 package io.micronaut.docs.replaces;
 
 import io.micronaut.context.ApplicationContext;
+import io.micronaut.docs.requires.Book;
 import org.junit.Assert;
 import org.junit.Test;
 
@@ -33,6 +34,8 @@ public class ReplacesTest {
         Assert.assertTrue(
                 applicationContext.getBean(BookService.class) instanceof MockBookService
         );
+        Assert.assertEquals("An OK Novel", applicationContext.getBean(Book.class).getTitle());
+        Assert.assertEquals("Learning 305", applicationContext.getBean(TextBook.class).getTitle());
 
         applicationContext.stop();
     }

--- a/inject-java/src/test/java/io/micronaut/docs/replaces/TextBook.java
+++ b/inject-java/src/test/java/io/micronaut/docs/replaces/TextBook.java
@@ -1,0 +1,10 @@
+package io.micronaut.docs.replaces;
+
+import io.micronaut.docs.requires.Book;
+
+public class TextBook extends Book {
+
+    public TextBook(String title) {
+        super(title);
+    }
+}

--- a/inject-java/src/test/java/io/micronaut/docs/replaces/TextBookFactory.java
+++ b/inject-java/src/test/java/io/micronaut/docs/replaces/TextBookFactory.java
@@ -1,0 +1,18 @@
+package io.micronaut.docs.replaces;
+
+import io.micronaut.context.annotation.Factory;
+import io.micronaut.context.annotation.Replaces;
+
+import javax.inject.Singleton;
+
+// tag::class[]
+@Factory
+public class TextBookFactory {
+
+    @Singleton
+    @Replaces(value = TextBook.class, factory = BookFactory.class)
+    TextBook textBook() {
+        return new TextBook("Learning 305");
+    }
+}
+// end::class[]

--- a/inject/src/main/java/io/micronaut/context/AbstractBeanDefinition.java
+++ b/inject/src/main/java/io/micronaut/context/AbstractBeanDefinition.java
@@ -308,8 +308,8 @@ public class AbstractBeanDefinition<T> extends AbstractBeanContextConditional im
     }
 
     @Override
-    public Class<?> getDeclaringType() {
-        return declaringType;
+    public Optional<Class<?>> getDeclaringType() {
+        return Optional.ofNullable(declaringType);
     }
 
     @Override

--- a/inject/src/main/java/io/micronaut/context/AbstractBeanDefinition.java
+++ b/inject/src/main/java/io/micronaut/context/AbstractBeanDefinition.java
@@ -308,6 +308,11 @@ public class AbstractBeanDefinition<T> extends AbstractBeanContextConditional im
     }
 
     @Override
+    public Class<?> getDeclaringType() {
+        return declaringType;
+    }
+
+    @Override
     public ConstructorInjectionPoint<T> getConstructor() {
         return constructor;
     }

--- a/inject/src/main/java/io/micronaut/context/BeanDefinitionDelegate.java
+++ b/inject/src/main/java/io/micronaut/context/BeanDefinitionDelegate.java
@@ -103,7 +103,7 @@ class BeanDefinitionDelegate<T> extends AbstractBeanContextConditional implement
     }
 
     @Override
-    public Class<?> getDeclaringType() {
+    public Optional<Class<?>> getDeclaringType() {
         return definition.getDeclaringType();
     }
 

--- a/inject/src/main/java/io/micronaut/context/BeanDefinitionDelegate.java
+++ b/inject/src/main/java/io/micronaut/context/BeanDefinitionDelegate.java
@@ -103,6 +103,11 @@ class BeanDefinitionDelegate<T> extends AbstractBeanContextConditional implement
     }
 
     @Override
+    public Class<?> getDeclaringType() {
+        return definition.getDeclaringType();
+    }
+
+    @Override
     public ConstructorInjectionPoint<T> getConstructor() {
         return definition.getConstructor();
     }

--- a/inject/src/main/java/io/micronaut/context/DefaultBeanContext.java
+++ b/inject/src/main/java/io/micronaut/context/DefaultBeanContext.java
@@ -26,6 +26,7 @@ import io.micronaut.context.scope.CustomScope;
 import io.micronaut.context.scope.CustomScopeRegistry;
 import io.micronaut.core.annotation.AnnotationMetadata;
 import io.micronaut.core.annotation.AnnotationMetadataProvider;
+import io.micronaut.core.annotation.AnnotationValue;
 import io.micronaut.core.async.subscriber.Completable;
 import io.micronaut.core.convert.ConversionService;
 import io.micronaut.core.io.ResourceLoader;
@@ -1353,46 +1354,71 @@ public class DefaultBeanContext implements BeanContext {
     }
 
     private <T> void filterReplacedBeans(Collection<BeanDefinition<T>> candidates) {
-        List<Class> replacedTypes = new ArrayList<>(2);
-        for (BeanDefinition<T> candidate : candidates) {
-            Optional<Class> replacesType = candidate.getAnnotationMetadata().getValue(Replaces.class, Class.class);
+        List<AnnotationValue<Replaces>> replacedTypes = new ArrayList<>(2);
 
-            replacesType.ifPresent(aClass -> {
+        for (BeanDefinition<T> candidate : candidates) {
+            Optional<AnnotationValue<Replaces>> replaces = candidate.findAnnotation(Replaces.class);
+
+            replaces.ifPresent(r -> {
                 if (LOG.isDebugEnabled()) {
-                    LOG.debug("Bean [{}] replaces existing bean of type [{}]", candidate.getBeanType(), aClass);
+                    Optional<Class> type = r.getValue(Class.class);
+                    Optional<Class> factory = r.get("factory", Class.class);
+                    if (factory.isPresent()) {
+                        LOG.debug("Bean [{}] replaces existing bean of type [{}] in factory type [{}]", candidate.getBeanType(), type.orElse(null), factory.get());
+                    } else {
+                        LOG.debug("Bean [{}] replaces existing bean of type [{}]", candidate.getBeanType(), type.orElse(null));
+                    }
                 }
-                replacedTypes.add(aClass);
+                replacedTypes.add(r);
             });
         }
         if (!replacedTypes.isEmpty()) {
+
             candidates.removeIf(definition -> {
-                        if (definition.hasDeclaredStereotype(Infrastructure.class)) {
+                if (definition.hasDeclaredStereotype(Infrastructure.class)) {
+                    return false;
+                }
+
+                Class<?> declaringType = definition.getDeclaringType();
+                Function<Class, Boolean> comparisonFunction = typeMatches(definition);
+
+                return replacedTypes.stream().anyMatch(r -> {
+                    Optional<Class> factory = r.get("factory", Class.class);
+                    Optional<Class> beanType = r.getValue(Class.class);
+                    if (factory.isPresent()) {
+                        if (factory.get() == declaringType) {
+                            return !beanType.isPresent() || comparisonFunction.apply(beanType.get());
+                        } else {
                             return false;
                         }
-
-                        Class<T> bt = definition.getBeanType();
-                        if (definition.hasStereotype(INTRODUCTION_TYPE)) {
-                            Class<? super T> superclass = bt.getSuperclass();
-                            if (superclass == Object.class) {
-                                // interface introduction
-                                return replacedTypes.stream().anyMatch(t -> t.isAssignableFrom(bt));
-                            } else {
-                                // abstract class introduction
-                                return replacedTypes.contains(superclass);
-                            }
-
-                        } else if (definition.hasStereotype(AROUND_TYPE)) {
-                            Class<? super T> superclass = bt.getSuperclass();
-                            return replacedTypes.contains(superclass) || replacedTypes.contains(bt);
-                        } else {
-                            return replacedTypes.contains(bt);
-                        }
-
+                    } else {
+                        return beanType.map(comparisonFunction).orElse(false);
                     }
-            );
+                });
+            });
         }
     }
 
+    private <T> Function<Class, Boolean> typeMatches(BeanDefinition<T> definition) {
+        Class<T> bt = definition.getBeanType();
+
+        if (definition.hasStereotype(INTRODUCTION_TYPE)) {
+            Class<? super T> superclass = bt.getSuperclass();
+            if (superclass == Object.class) {
+                // interface introduction
+                return (clazz) -> clazz.isAssignableFrom(bt);
+            } else {
+                // abstract class introduction
+                return (clazz) -> clazz == superclass;
+            }
+        }
+        if (definition.hasStereotype(AROUND_TYPE)) {
+            Class<? super T> superclass = bt.getSuperclass();
+            return (clazz) -> clazz == superclass || clazz == bt;
+        }
+
+        return (clazz) -> clazz == bt;
+    }
 
     private <T> void doInject(BeanResolutionContext resolutionContext, T instance, BeanDefinition definition) {
         definition.inject(resolutionContext, this, instance);
@@ -2322,6 +2348,11 @@ public class DefaultBeanContext implements BeanContext {
         @Override
         public Class getBeanType() {
             return singletonClass;
+        }
+
+        @Override
+        public Class<?> getDeclaringType() {
+            return null;
         }
 
         @Override

--- a/inject/src/main/java/io/micronaut/context/DefaultBeanContext.java
+++ b/inject/src/main/java/io/micronaut/context/DefaultBeanContext.java
@@ -1379,14 +1379,14 @@ public class DefaultBeanContext implements BeanContext {
                     return false;
                 }
 
-                Class<?> declaringType = definition.getDeclaringType();
+                Optional<Class<?>> declaringType = definition.getDeclaringType();
                 Function<Class, Boolean> comparisonFunction = typeMatches(definition);
 
                 return replacedTypes.stream().anyMatch(r -> {
                     Optional<Class> factory = r.get("factory", Class.class);
                     Optional<Class> beanType = r.getValue(Class.class);
-                    if (factory.isPresent()) {
-                        if (factory.get() == declaringType) {
+                    if (factory.isPresent() && declaringType.isPresent()) {
+                        if (factory.get() == declaringType.get()) {
                             return !beanType.isPresent() || comparisonFunction.apply(beanType.get());
                         } else {
                             return false;
@@ -2351,8 +2351,8 @@ public class DefaultBeanContext implements BeanContext {
         }
 
         @Override
-        public Class<?> getDeclaringType() {
-            return null;
+        public Optional<Class<?>> getDeclaringType() {
+            return Optional.empty();
         }
 
         @Override

--- a/inject/src/main/java/io/micronaut/context/annotation/Replaces.java
+++ b/inject/src/main/java/io/micronaut/context/annotation/Replaces.java
@@ -35,5 +35,10 @@ public @interface Replaces {
     /**
      * @return The bean type that this bean replaces
      */
-    Class value();
+    Class value() default void.class;
+
+    /**
+     * @return The declaring bean type
+     */
+    Class factory() default void.class;
 }

--- a/inject/src/main/java/io/micronaut/inject/BeanDefinition.java
+++ b/inject/src/main/java/io/micronaut/inject/BeanDefinition.java
@@ -72,7 +72,7 @@ public interface BeanDefinition<T> extends AnnotationMetadataDelegate, Named, Be
     /**
      * @return The type that declares this definition, null if not applicable.
      */
-    Class<?> getDeclaringType();
+    Optional<Class<?>> getDeclaringType();
 
     /**
      * The single concrete constructor that is an injection point for creating the bean.

--- a/inject/src/main/java/io/micronaut/inject/BeanDefinition.java
+++ b/inject/src/main/java/io/micronaut/inject/BeanDefinition.java
@@ -70,6 +70,11 @@ public interface BeanDefinition<T> extends AnnotationMetadataDelegate, Named, Be
     Class<T> getBeanType();
 
     /**
+     * @return The type that declares this definition, null if not applicable.
+     */
+    Class<?> getDeclaringType();
+
+    /**
      * The single concrete constructor that is an injection point for creating the bean.
      *
      * @return The constructor injection point

--- a/src/main/docs/guide/ioc/replaces.adoc
+++ b/src/main/docs/guide/ioc/replaces.adoc
@@ -29,4 +29,34 @@ include::{testsjava}/replaces/MockBookService.java[tags=class, indent=0]
 ----
 
 <1> The `MockBookService` declares that it replaces `JdbcBookService`
-<2> But only for for the `test` environment
+
+The `@Replaces` annotation also supports a `factory` argument. That argument allows the replacement of factory beans in their entirety or specific types created by the factory.
+
+For example, it may be desired to replace all or part of the given factory class:
+
+.BookFactory.java
+[source,java]
+----
+include::{testsjava}/replaces/BookFactory.java[tags=class, indent=0]
+----
+
+WARNING: To replace a factory in its entirety, it is necessary that your factory methods match the return types of all of the methods in the replaced factory.
+
+In this example, the `BookFactory#textBook()` will *not* be replaced because this factory does not have a factory method that returns a TextBook.
+
+.CustomBookFactory.java
+[source,java]
+----
+include::{testsjava}/replaces/CustomBookFactory.java[tags=class, indent=0]
+----
+
+It may be the case that you don't wish for the factory methods to be replaced, except for a select few. For that use case, you can apply the `@Replaces` annotation on the method and denote the factory that it should apply to.
+
+.TextBookFactory.java
+[source,java]
+----
+include::{testsjava}/replaces/TextBookFactory.java[tags=class, indent=0]
+----
+
+The `BookFactory#novel()` method will not be replaced because the TextBook class is defined in the annotation.
+

